python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,201 @@
+                # new leaf
+                preimage = nibble_path_to_bytes(path + r_node.rest_of_key)
+                address = self._address_preimages[preimage]
+                self._main_trie[address] = tuple((None, AccountNode.from_rlp(r_node.value)))
+
+            case "None", "ExtensionNode":
+                r_node: ExtensionNode = cast(ExtensionNode, r_node)
+                # Look for diffs in the right sub-tree
+                self._compute_diff(None, r_node.subnode, path + r_node.key_segment)
+
+            case "None", "BranchNode":
+                # Look for diffs in all branches of the right sub-tree
+                r_node: BranchNode = cast(BranchNode, r_node)
+                for i in range(0, 16):
+                    self._compute_diff(None, r_node.subnodes[i], path + i.to_bytes(1, "big"))
+
+            case "LeafNode", "None":
+                # deleted leaf (should not happen post-cancun)
+                preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                address = self._address_preimages[preimage]
+                self._main_trie[address] = tuple((AccountNode.from_rlp(l_node.value), None))
+
+            case "LeafNode", "LeafNode":
+                if l_node.value != r_node.value:
+                    preimage = nibble_path_to_bytes(path + l_node.rest_of_key)
+                    address = self._address_preimages[preimage]
+                    self._main_trie[address] = tuple((AccountNode.from_rlp(l_node.value), AccountNode.from_rlp(r_node.value)))
+
+            case "LeafNode", "BranchNode":
+                # Left is a valute, right is
+                logger.warning("LeafNode -> BranchNode")
+                breakpoint()
+                for i in range(0, 16):
+                    self._compute_diff(None, r_node.subnodes[i], path + i.to_bytes(1, "big"))
+
+            case "LeafNode", "ExtensionNode":
+                raise ValueError("LeafNode -> ExtensionNode should not happen if comparing two valid trie roots")
+
+            case "ExtensionNode", "None":
+                # Look for diffs in the left sub-tree
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                self._compute_diff(l_node.subnode, None, path + l_node.key_segment)
+
+            case "ExtensionNode", "LeafNode":
+                raise ValueError("ExtensionNode -> LeafNode should not happen if comparing two valid trie roots")
+
+            case "ExtensionNode", "ExtensionNode":
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                r_node: ExtensionNode = cast(ExtensionNode, r_node)
+
+                # Equal keys -> Look for diffs in childrens
+                if l_node.key_segment == r_node.key_segment:
+                    self._compute_diff(l_node.subnode, r_node.subnode, path + l_node.key_segment)
+
+                # Right is prefix of left -> Look for diffs in left sub-tree
+                elif l_node.key_segment.startswith(r_node.key_segment):
```

Comment:
why do we stop exploring the right trie?
in my understanding, if it is a prefix of left, then it means a branch node was inserted at the divergence point between l and r, which suggests there is a diff inside of it too